examples/mnist_denoising_autoencoder.py

```diff
@@ -0,0 +1,142 @@
+'''Trains a denoising autoenconder on MNIST dataset.
+
+Denoising is one of the classic applications of autoencoders.
+The denoising process removes unwanted noise that corrupted the
+true signal.
+
+Noise + Data ---> Denoising Autoencoder ---> Data
+
+Given a training dataset of corrupted data as input and
+true signal as output, a denoising autoencoder can recover the
+hidden structure to generate clean data.
+
+This example has modular design. The encoder, decoder and autoencoder
+are 3 models that share weights. For example, after training the
+autoencoder, the encoder can be used to  generate latent vectors
+of input data for low-dim visualization like PCA or TSNE.
+'''
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+import keras
+from keras.layers import Activation, Dense, Input
+from keras.layers import Conv2D, Flatten
+from keras.layers import Reshape, Conv2DTranspose
+from keras.models import Model
+from keras import backend as K
+from keras.datasets import mnist
+import numpy as np
+import matplotlib.pyplot as plt
+
+# MNIST dataset
+(x_train, _), (x_test, _) = mnist.load_data()
+
+image_size = x_train.shape[1]
+x_train = np.reshape(x_train, [-1, image_size, image_size, 1])
+x_test = np.reshape(x_test, [-1, image_size, image_size, 1])
+x_train = x_train.astype('float32') / 255
+x_test = x_test.astype('float32') / 255
+
+# Generate corrupted MNIST images by adding noise with normal dist
+# centered at 0.5 and std=0.5
+noise = np.random.normal(loc=0.5, scale=0.5, size=x_train.shape)
+x_train_noisy = x_train + noise
+noise = np.random.normal(loc=0.5, scale=0.5, size=x_test.shape)
+x_test_noisy = x_test + noise
+
+x_train_noisy = np.clip(x_train_noisy, 0., 1.)
+x_test_noisy = np.clip(x_test_noisy, 0., 1.)
+
+# Network parameters
+input_shape = (image_size, image_size, 1)
+batch_size = 128
+kernel_size = 3
+pool_size = 2
+filters = 16
+latent_dim = 16
+
+# Build the Autoencoder Model
+# First build the Encoder Model
+inputs = Input(shape=input_shape, name='encoder_input')
+x = inputs
+# Stack of Conv2D blocks
+# Notes:
+# 1) Use Batch Normalization before ReLU on deep networks
+# 2) Use MaxPooling2D as alternative to strides>1
+# - faster but not as good as strides>1
+for i in range(2):
```

Comment:
Here we actually start with 32 filters even though we previously defined `filters = 16`. What would you think of instead having:

```python
layer_filters = [32, 64]
for filters in layer_filters:
    ...

for filters in layer_filters[::-1]:
   # decoder
   ...
```
This allows arbitrary layer size and arbitrary number of layers.

Reply:
In encoder, the doubling of filter size was actually inspired by autoencoders used in practice, by ResNet and by DenseNet (that uses growth rate). As the number of feature maps is halved, the number of filters doubles. 
In decoder, the halving of filter size has the same motivation. 
Use of array is another way to do it. More flexible but requires more care due to manual encoding of hyperparameters.

Reply:
You are manually encoding the hyperparameters in both cases (number of layers and size of each layer).

Reply:
I Agree. Use of list results to a more compact and flexible code. The autoencoder model structure, like depth, is not constrained to halving/doubling of number of filters limitations. I submitted the revised code.